### ISSUE TITLE
5.4.119 Import fails with NullReferenceException

### ISSUE THREAD

**Issue by dusk0r** [ravendb/ravendb#18635]
Exporting and importing a collection with attachments causes sometimes following exceptions:

```
[13:05:07 INFO] Started processing Documents.
Object reference not set to an instance of an object.
System.NullReferenceException: Object reference not set to an instance of an object.
   at Sparrow.Json.BlittableJsonReaderObject.GetPropertiesByInsertionOrder() in D:\Builds\RavenDB-Stable-5.4\54129\src\Sparrow\Json\BlittableJsonReaderObject.cs:line 917
   at Sparrow.Json.Parsing.ObjectJsonParser.Read() in D:\Builds\RavenDB-Stable-5.4\54129\src\Sparrow\Json\Parsing\ObjectJsonParser.cs:line 244
   at Sparrow.Json.BlittableJsonDocumentBuilder.ReadInternal[TWriteStrategy]() in D:\Builds\RavenDB-Stable-5.4\54129\src\Sparrow\Json\BlittableJsonDocumentBuilder.cs:line 118
   at Sparrow.Json.JsonOperationContext.ReadObjectInternal(Object builder, String documentId, UsageMode mode, IBlittableDocumentModifier modifier) in D:\Builds\RavenDB-Stable-5.4\54129\src\Sparrow\Json\JsonOperationContext.cs:line 577
   at Raven.Server.Documents.DocumentPutAction.Recreate(DocumentsOperationContext context, String id, BlittableJsonReaderObject oldDoc, BlittableJsonReaderObject& document, DocumentFlags& flags, NonPersistentDocumentFlags nonPersistentFlags, UInt64& documentDebugHash) in D:\Builds\RavenDB-Stable-5.4\54129\src\Raven.Server\Documents\DocumentPutAction.cs:line 471
   at Raven.Server.Documents.DocumentPutAction.PutDocument(DocumentsOperationContext context, String id, String expectedChangeVector, BlittableJsonReaderObject document, Nullable`1 lastModifiedTicks, String changeVector, String oldChangeVectorForClusterTransactionIndexCheck, DocumentFlags flags, NonPersistentDocumentFlags nonPersistentFlags) in D:\Builds\RavenDB-Stable-5.4\54129\src\Raven.Server\Documents\DocumentPutAction.cs:line 314
   at Raven.Server.Smuggler.Documents.DatabaseDestination.MergedBatchPutCommand.ExecuteCmd(DocumentsOperationContext context) in D:\Builds\RavenDB-Stable-5.4\54129\src\Raven.Server\Smuggler\Documents\DatabaseDestination.cs:line 1714
   at Raven.Server.Documents.TransactionOperationsMerger.ExecutePendingOperationsInTransaction(List`1 executedOps, DocumentsOperationContext context, Task previousOperation, DurationMeasurement& meter) in D:\Builds\RavenDB-Stable-5.4\54129\src\Raven.Server\Documents\TransactionOperationsMerger.cs:line 930
   at Raven.Server.Documents.TransactionOperationsMerger.MergeTransactionsOnce() in D:\Builds\RavenDB-Stable-5.4\54129\src\Raven.Server\Documents\TransactionOperationsMerger.cs:line 517
--- End of stack trace from previous location ---
   at Raven.Server.Documents.TransactionOperationsMerger.Enqueue(MergedTransactionCommand cmd) in D:\Builds\RavenDB-Stable-5.4\54129\src\Raven.Server\Documents\TransactionOperationsMerger.cs:line 142
   at Raven.Server.Smuggler.Documents.DatabaseDestination.DatabaseDocumentActions.<>c__DisplayClass34_0.<<HandleBatchOfDocumentsIfNecessaryAsync>g__SubmitNextBatch|0>d.MoveNext() in D:\Builds\RavenDB-Stable-5.4\54129\src\Raven.Server\Smuggler\Documents\DatabaseDestination.cs:line 565
--- End of stack trace from previous location ---
   at Raven.Server.Smuggler.Documents.DatabaseSmuggler.ProcessDocumentsAsync(SmugglerResult result, BuildVersionType buildType) in D:\Builds\RavenDB-Stable-5.4\54129\src\Raven.Server\Smuggler\Documents\DatabaseSmuggler.cs:line 793
   at Raven.Server.Smuggler.Documents.DatabaseSmuggler.ProcessDocumentsAsync(SmugglerResult result, BuildVersionType buildType) in D:\Builds\RavenDB-Stable-5.4\54129\src\Raven.Server\Smuggler\Documents\DatabaseSmuggler.cs:line 793
   at Raven.Server.Smuggler.Documents.DatabaseSmuggler.ProcessDocumentsAsync(SmugglerResult result, BuildVersionType buildType) in D:\Builds\RavenDB-Stable-5.4\54129\src\Raven.Server\Smuggler\Documents\DatabaseSmuggler.cs:line 793
   at Raven.Server.Smuggler.Documents.DatabaseSmuggler.ProcessDocumentsAsync(SmugglerResult result, BuildVersionType buildType) in D:\Builds\RavenDB-Stable-5.4\54129\src\Raven.Server\Smuggler\Documents\DatabaseSmuggler.cs:line 793
   at Raven.Server.Smuggler.Documents.DatabaseSmuggler.ProcessTypeAsync(DatabaseItemType type, SmugglerResult result, BuildVersionType buildType, Boolean ensureStepsProcessed) in D:\Builds\RavenDB-Stable-5.4\54129\src\Raven.Server\Smuggler\Documents\DatabaseSmuggler.cs:line 292
   at Raven.Server.Smuggler.Documents.DatabaseSmuggler.ExecuteAsync(Boolean ensureStepsProcessed, Boolean isLastFile) in D:\Builds\RavenDB-Stable-5.4\54129\src\Raven.Server\Smuggler\Documents\DatabaseSmuggler.cs:line 111
   at Raven.Server.Smuggler.Documents.DatabaseSmuggler.ExecuteAsync(Boolean ensureStepsProcessed, Boolean isLastFile) in D:\Builds\RavenDB-Stable-5.4\54129\src\Raven.Server\Smuggler\Documents\DatabaseSmuggler.cs:line 111
   at Raven.Server.Smuggler.Documents.Handlers.SmugglerHandler.DoImportInternalAsync(DocumentsOperationContext context, Stream stream, DatabaseSmugglerOptionsServerSide options, SmugglerResult result, Action`1 onProgress, OperationCancelToken token) in D:\Builds\RavenDB-Stable-5.4\54129\src\Raven.Server\Smuggler\Documents\Handlers\SmugglerHandler.cs:line 962
   at Raven.Server.Smuggler.Documents.Handlers.SmugglerHandler.DoImportInternalAsync(DocumentsOperationContext context, Stream stream, DatabaseSmugglerOptionsServerSide options, SmugglerResult result, Action`1 onProgress, OperationCancelToken token) in D:\Builds\RavenDB-Stable-5.4\54129\src\Raven.Server\Smuggler\Documents\Handlers\SmugglerHandler.cs:line 962
   at Raven.Server.Smuggler.Documents.Handlers.SmugglerHandler.<>c__DisplayClass16_1.<<PostImportAsync>b__1>d.MoveNext() in D:\Builds\RavenDB-Stable-5.4\54129\src\Raven.Server\Smuggler\Documents\Handlers\SmugglerHandler.cs:line 735
```

```
[07:57:31 INFO] Started processing DatabaseRecord.
Index was outside the bounds of the array.
System.IndexOutOfRangeException: Index was outside the bounds of the array.
   at Sparrow.Json.Parsing.ObjectJsonParser.Read() in D:\Builds\RavenDB-Stable-5.4\54129\src\Sparrow\Json\Parsing\ObjectJsonParser.cs:line 244
   at Sparrow.Json.BlittableJsonDocumentBuilder.ReadInternal[TWriteStrategy]() in D:\Builds\RavenDB-Stable-5.4\54129\src\Sparrow\Json\BlittableJsonDocumentBuilder.cs:line 118
   at Sparrow.Json.JsonOperationContext.ReadObjectInternal(Object builder, String documentId, UsageMode mode, IBlittableDocumentModifier modifier) in D:\Builds\RavenDB-Stable-5.4\54129\src\Sparrow\Json\JsonOperationContext.cs:line 577
   at Raven.Server.Documents.DocumentPutAction.Recreate(DocumentsOperationContext context, String id, BlittableJsonReaderObject oldDoc, BlittableJsonReaderObject& document, DocumentFlags& flags, NonPersistentDocumentFlags nonPersistentFlags, UInt64& documentDebugHash) in D:\Builds\RavenDB-Stable-5.4\54129\src\Raven.Server\Documents\DocumentPutAction.cs:line 471
   at Raven.Server.Documents.DocumentPutAction.PutDocument(DocumentsOperationContext context, String id, String expectedChangeVector, BlittableJsonReaderObject document, Nullable`1 lastModifiedTicks, String changeVector, String oldChangeVectorForClusterTransactionIndexCheck, DocumentFlags flags, NonPersistentDocumentFlags nonPersistentFlags) in D:\Builds\RavenDB-Stable-5.4\54129\src\Raven.Server\Documents\DocumentPutAction.cs:line 314
   at Raven.Server.Smuggler.Documents.DatabaseDestination.MergedBatchPutCommand.ExecuteCmd(DocumentsOperationContext context) in D:\Builds\RavenDB-Stable-5.4\54129\src\Raven.Server\Smuggler\Documents\DatabaseDestination.cs:line 1714
   at Raven.Server.Documents.TransactionOperationsMerger.ExecutePendingOperationsInTransaction(List`1 executedOps, DocumentsOperationContext context, Task previousOperation, DurationMeasurement& meter) in D:\Builds\RavenDB-Stable-5.4\54129\src\Raven.Server\Documents\TransactionOperationsMerger.cs:line 930
   at Raven.Server.Documents.TransactionOperationsMerger.MergeTransactionsOnce() in D:\Builds\RavenDB-Stable-5.4\54129\src\Raven.Server\Documents\TransactionOperationsMerger.cs:line 517
--- End of stack trace from previous location ---
   at Raven.Server.Documents.TransactionOperationsMerger.Enqueue(MergedTransactionCommand cmd) in D:\Builds\RavenDB-Stable-5.4\54129\src\Raven.Server\Documents\TransactionOperationsMerger.cs:line 142
   at Raven.Server.Smuggler.Documents.DatabaseDestination.DatabaseDocumentActions.<>c__DisplayClass34_0.<<HandleBatchOfDocumentsIfNecessaryAsync>g__SubmitNextBatch|0>d.MoveNext() in D:\Builds\RavenDB-Stable-5.4\54129\src\Raven.Server\Smuggler\Documents\DatabaseDestination.cs:line 565
--- End of stack trace from previous location ---
   at Raven.Server.Smuggler.Documents.DatabaseSmuggler.ProcessDocumentsAsync(SmugglerResult result, BuildVersionType buildType) in D:\Builds\RavenDB-Stable-5.4\54129\src\Raven.Server\Smuggler\Documents\DatabaseSmuggler.cs:line 793
   at Raven.Server.Smuggler.Documents.DatabaseSmuggler.ProcessDocumentsAsync(SmugglerResult result, BuildVersionType buildType) in D:\Builds\RavenDB-Stable-5.4\54129\src\Raven.Server\Smuggler\Documents\DatabaseSmuggler.cs:line 793
   at Raven.Server.Smuggler.Documents.DatabaseSmuggler.ProcessDocumentsAsync(SmugglerResult result, BuildVersionType buildType) in D:\Builds\RavenDB-Stable-5.4\54129\src\Raven.Server\Smuggler\Documents\DatabaseSmuggler.cs:line 793
   at Raven.Server.Smuggler.Documents.DatabaseSmuggler.ProcessDocumentsAsync(SmugglerResult result, BuildVersionType buildType) in D:\Builds\RavenDB-Stable-5.4\54129\src\Raven.Server\Smuggler\Documents\DatabaseSmuggler.cs:line 793
   at Raven.Server.Smuggler.Documents.DatabaseSmuggler.ProcessTypeAsync(DatabaseItemType type, SmugglerResult result, BuildVersionType buildType, Boolean ensureStepsProcessed) in D:\Builds\RavenDB-Stable-5.4\54129\src\Raven.Server\Smuggler\Documents\DatabaseSmuggler.cs:line 292
   at Raven.Server.Smuggler.Documents.DatabaseSmuggler.ExecuteAsync(Boolean ensureStepsProcessed, Boolean isLastFile) in D:\Builds\RavenDB-Stable-5.4\54129\src\Raven.Server\Smuggler\Documents\DatabaseSmuggler.cs:line 111
   at Raven.Server.Smuggler.Documents.DatabaseSmuggler.ExecuteAsync(Boolean ensureStepsProcessed, Boolean isLastFile) in D:\Builds\RavenDB-Stable-5.4\54129\src\Raven.Server\Smuggler\Documents\DatabaseSmuggler.cs:line 111
   at Raven.Server.Smuggler.Documents.Handlers.SmugglerHandler.DoImportInternalAsync(DocumentsOperationContext context, Stream stream, DatabaseSmugglerOptionsServerSide options, SmugglerResult result, Action`1 onProgress, OperationCancelToken token) in D:\Builds\RavenDB-Stable-5.4\54129\src\Raven.Server\Smuggler\Documents\Handlers\SmugglerHandler.cs:line 962
   at Raven.Server.Smuggler.Documents.Handlers.SmugglerHandler.DoImportInternalAsync(DocumentsOperationContext context, Stream stream, DatabaseSmugglerOptionsServerSide options, SmugglerResult result, Action`1 onProgress, OperationCancelToken token) in D:\Builds\RavenDB-Stable-5.4\54129\src\Raven.Server\Smuggler\Documents\Handlers\SmugglerHandler.cs:line 962
   at Raven.Server.Smuggler.Documents.Handlers.SmugglerHandler.<>c__DisplayClass16_1.<<PostImportAsync>b__1>d.MoveNext() in D:\Builds\RavenDB-Stable-5.4\54129\src\Raven.Server\Smuggler\Documents\Handlers\SmugglerHandler.cs:line 735
--- End of stack trace from previous location ---
   at Raven.Server.Smuggler.Documents.Handlers.SmugglerHandler.PostImportAsync() in D:\Builds\RavenDB-Stable-5.4\54129\src\Raven.Server\Smuggler\Documents\Handlers\SmugglerHandler.cs:line 742
```

This happens with around every second import. 
The exceptions only occure if attachments are present, even when the attachments aren't imported. Exports without attachments will import without problems.
We tested with multiple computers and RavenDB Versions between 5.4.112 to 5.4.119 on Windows11. 

I can provide you a .ravendbdump in private.

**Comment by grisha-kotler** [comment]
Hi @dusk0r,
Can you share the .ravendbdump file? (to support@ravendb.net)

**Comment by dusk0r** [comment]
I send a mail. Hope you can reproduce.

**Comment by dusk0r** [comment]
Hi @grisha-kotler,
Any luck replicating the issue? Can we provide any help?

**Comment by grisha-kotler** [comment]
@dusk0r, you can track the status of this issue here:
https://issues.hibernatingrhinos.com/issue/RavenDB-22046/Register-dispose-for-builder-modifier-to-MergedBatchPutCommand

**Comment by dusk0r** [comment]
Fixed in 5.4.201 / 6.0.105
